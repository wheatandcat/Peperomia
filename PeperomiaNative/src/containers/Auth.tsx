import * as GoogleSignIn from "expo-google-sign-in";
import * as Google from "expo-google-app-auth";
import { AsyncStorage, Platform } from "react-native";
import React, { createContext, Component } from "react";
import Constants from "expo-constants";
import { Sentry } from "react-native-sentry";
import * as firebase from "firebase";

const Context = createContext({});
const { Provider } = Context;

interface Props {}

interface State {
  uid: string;
  email: string;
}

const isStandaloneAndAndroid = () => {
  return Platform.OS === "android" && Constants.appOwnership !== "expo";
};

export default class extends Component<Props, State> {
  state = {
    email: "",
    uid: ""
  };

  async componentDidMount() {
    if (isStandaloneAndAndroid()) {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      try {
        await GoogleSignIn.initAsync({
          clientId: String(androidClientId)
        });
      } catch ({ message }) {
        Sentry.captureMessage(JSON.stringify(message));
      }
    }

    const loggedIn = await this.loggedIn();

    if (loggedIn && !this.state.uid) {
      const uid = await AsyncStorage.getItem("uid");
      if (uid) {
        this.setState({
          uid
        });
      }
    }

    if (loggedIn && !this.state.email) {
      const email = await AsyncStorage.getItem("email");
      if (email) {
        this.setState({
          email
        });
      }
    }
  }

  onGoogleLogin = async () => {
    if (isStandaloneAndAndroid()) {
      // TODO: AndroidのstandaloneのみGoogleSignInを使わないとエラーになる
      // https://github.com/expo/expo/issues/4762

      await GoogleSignIn.askForPlayServicesAsync();
      const result = await GoogleSignIn.signInAsync();

      if (result.type === "success" && result.user && result.user.auth) {
        const { idToken, accessToken } = result.user.auth;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    } else {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      const iosClientId = process.env.GOOGLE_LOGIN_IOS_CLIENT_ID;
      const result = await Google.logInAsync({
        clientId:
          Platform.OS === "ios" ? String(iosClientId) : String(androidClientId),
        iosClientId,
        androidClientId,
        scopes: ["profile", "email"]
      });

      if (result.type === "success") {
        const { idToken, accessToken } = result;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    }
  };

  firebaseLogin = async (idToken: string, accessToken: string) => {
    const credential = firebase.auth.GoogleAuthProvider.credential(
      idToken,
      accessToken
    );

    await firebase
      .auth()
      .signInAndRetrieveDataWithCredential(credential)
      .catch(error => {
        Sentry.captureMessage(JSON.stringify(error));
      });

    await this.setSession(true);
  };

  loggedIn = async () => {
    const idToken = await this.getIdToken();

    return Boolean(idToken);
  };

  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem("id_token");
    await AsyncStorage.removeItem("expiration");
    await AsyncStorage.removeItem("email");
  };

  setSession = async (refresh = false) => {
    const user = firebase.auth().currentUser;
    if (!user) {
      return;
    }

    if (user.email) {
      await AsyncStorage.setItem("email", user.email);
      await AsyncStorage.setItem("uid", user.uid);
      this.setState({
        email: user.email,
        uid: user.uid
      });
    }

    const idToken = await user.getIdToken(refresh);
    await AsyncStorage.setItem("id_token", idToken);
    await AsyncStorage.setItem(
      "expiration",
      String(new Date().getTime() + 60 * 60)
    );

    return idToken;
  };

  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem("id_token");
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem("expiration");
    if (Number(expiration) > new Date().getTime()) {
      return idToken;
    }

    return this.setSession(true);
  };

  render() {
    return (
      <Provider
        value={{
          onGoogleLogin: this.onGoogleLogin,
          getIdToken: this.getIdToken,
          loggedIn: this.loggedIn,
          logout: this.logout,
          email: this.state.email,
          uid: this.state.uid
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
