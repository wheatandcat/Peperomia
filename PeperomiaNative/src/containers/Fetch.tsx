import React, { createContext, Component, ReactNode } from 'react';
import Constants from 'expo-constants';
import {
  Consumer as AuthConsumer,
  ContextProps as AuthContextProps,
} from './Auth';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

export type ContextProps = Partial<Pick<Connected, 'post'>>;

type Props = {
  children: ReactNode;
};

export default class extends Component<Props> {
  render() {
    return (
      <AuthConsumer>
        {({ getIdToken }: AuthContextProps) => (
          <Connected getIdToken={getIdToken}>{this.props.children}</Connected>
        )}
      </AuthConsumer>
    );
  }
}

type ConnectedProps = Pick<AuthContextProps, 'getIdToken'>;

export interface FetchValue {
  post: (url: string, param: any) => Promise<Response>;
}

class Connected extends Component<ConnectedProps> {
  post = async (url: string, param: any): Promise<Response> => {
    const idToken = this.props.getIdToken ? await this.props.getIdToken() : '';
    const apiHost = process.env.API_HOST;
    const init = {
      ...param,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${idToken}`,
      },
    };

    if (!Constants.isDevice) {
      console.info(init);
    }

    console.info(`Bearer ${idToken}`);

    const response = await fetch(`${apiHost}/${url}`, init);
    console.log(response);
    if (!Constants.isDevice) {
      if (!response.ok) {
        const result = await response.json();
        console.info(result);
      }
    }

    return response;
  };

  render() {
    return (
      <Provider
        value={{
          post: this.post,
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
