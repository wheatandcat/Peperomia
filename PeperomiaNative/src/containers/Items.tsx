import { SQLite } from "expo-sqlite";
import React, { createContext, Component } from "react";
import { db } from "../lib/db";
import { select as selectItems, Item } from "../lib/db/item";
import {
  selectByItemId as selectItemDetailByItemId,
  ItemDetail
} from "../lib/db/itemDetail";

const Context = createContext({});
const { Provider } = Context;

interface ItemAbout {
  itemId: number;
  about: string;
}

interface Props {}

interface State {
  loading: boolean;
  items: Item[];
  about: ItemAbout[];
}

export default class extends Component<Props, State> {
  state = {
    loading: true,
    items: [],
    about: []
  };

  componentDidMount() {
    this.getData();
  }

  getData = () => {
    db.transaction((tx: SQLite.Transaction) => {
      selectItems(tx, this.setItems);
    });
  };

  setItems = (data: any, error: any) => {
    if (error) {
      return;
    }
    this.setState({
      items: data,
      about: []
    });

    data.map((val: Item) => {
      db.transaction((tx: SQLite.Transaction) => {
        selectItemDetailByItemId(tx, String(val.id), this.setItemsDetail);
      });
    });
  };

  setItemsDetail = (data: any, error: any) => {
    if (error || !data || data.length === 0) {
      return;
    }

    const names = data.map((val: ItemDetail) => val.title).join("â†’");
    const itemId = data[0].itemId;
    const about = [
      ...this.state.about,
      {
        itemId: itemId,
        about: names
      }
    ];

    this.setState({
      about,
      loading: false
    });
  };

  render() {
    return (
      <Provider
        value={{
          items: this.state.items,
          about: this.state.about,
          refreshData: this.getData,
          itemsLoading: this.state.loading
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
