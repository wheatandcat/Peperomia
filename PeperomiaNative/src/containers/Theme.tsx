import React, { createContext, Component } from "react";
import { AsyncStorage } from "react-native";
import Spinner from "react-native-loading-spinner-overlay";
import { setMode } from "../config/theme";

const Context = createContext({});
const { Provider } = Context;

type Props = {};

type State = {
  rerendering: boolean;
  mode: string;
  render: boolean;
};

export type ContextProps = Partial<
  Pick<State, "rerendering" | "mode"> & {
    onModeChange: (mode: "light" | "dark") => void;
    onFinishRerendering: () => void;
  }
>;

export default class extends Component<Props, State> {
  state = {
    rerendering: false,
    mode: "light",
    render: false
  };

  async componentDidMount() {
    let mode = await AsyncStorage.getItem("THEME_MODE");
    console.log(mode);

    if (mode !== "dark") {
      mode = "light";
    }

    if (mode === "light" || mode === "dark") {
      setMode(mode);
    }

    this.setState({
      mode,
      render: true
    });
  }

  onModeChange = async (mode: "light" | "dark") => {
    this.setState({
      mode,
      rerendering: true
    });

    await AsyncStorage.setItem("THEME_MODE", mode);

    if (mode === "light" || mode === "dark") {
      setMode(mode);
      setTimeout(this.setUnRender.bind(this), 100);
      setTimeout(this.setRender.bind(this), 101);
    }
  };

  setUnRender = () => {
    this.setState({
      render: false
    });
  };

  setRender = () => {
    this.setState({
      render: true
    });
  };

  onFinishRerendering = () => {
    this.setState({
      rerendering: false
    });
  };

  render() {
    return (
      <>
        <Spinner visible={this.state.rerendering} textContent="" />
        {this.state.render && (
          <Provider
            value={{
              rerendering: this.state.rerendering,
              mode: this.state.mode,
              onModeChange: this.onModeChange,
              onFinishRerendering: this.onFinishRerendering
            }}
          >
            {this.props.children}
          </Provider>
        )}
      </>
    );
  }
}

export const Consumer = Context.Consumer;
