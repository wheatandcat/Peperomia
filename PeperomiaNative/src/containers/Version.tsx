import { Component } from "react";
import { AsyncStorage } from "react-native";
import compareVersions from "compare-versions";
import { migrationV104 } from "../lib/migration";

interface Props {}

interface State {
  loading: boolean;
}

export default class extends Component<Props, State> {
  state = {
    loading: true
  };

  async componentDidMount() {
    let appVerion = await AsyncStorage.getItem("APP_VERSION");
    if (!appVerion) {
      appVerion = "1.0.0";
    }

    if (compareVersions.compare("1.0.4", appVerion, ">")) {
      // カラム追加のマイグレーション実行
      const ok = await migrationV104();
      if (ok) {
        // 現在のバージョンを設定
        AsyncStorage.setItem("APP_VERSION", "1.0.4");
      }
    }

    this.setState({
      loading: false
    });
  }

  render() {
    if (this.state.loading) {
      return null;
    }

    return this.props.children;
  }
}
