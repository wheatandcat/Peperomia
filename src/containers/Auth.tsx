import * as GoogleSignIn from 'expo-google-sign-in';
import * as Google from 'expo-google-app-auth';
import { Platform, Alert } from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import uuidv4 from 'uuid/v4';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import * as AppleAuthentication from 'expo-apple-authentication';
import React, {
  memo,
  createContext,
  FC,
  useState,
  useCallback,
  useContext,
  useEffect,
} from 'react';
import { View } from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import Spinner from 'react-native-loading-spinner-overlay';
import Constants from 'expo-constants';
import { backup } from 'lib/backup';
import * as Sentry from 'sentry-expo';
import firebase from 'lib/system/firebase';
import useIsFirstRender from 'hooks/useIsFirstRender';
import { UID } from 'domain/user';
import libAuth from 'lib/auth';
import {
  useSyncCalendarsMutation,
  SyncCalendarsMutationVariables,
  SyncCalendar,
  SyncItemDetail,
} from 'queries/api/index';
import { post } from 'lib/fetch';
import theme from 'config/theme';

dayjs.extend(advancedFormat);

const auth = new libAuth();

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type Props = {};

type State = {
  uid: UID;
  setup: boolean;
  render: boolean;
  email: string;
};

export type ContextProps = Partial<
  Pick<State, 'email' | 'uid'> & {
    onGoogleLogin: () => Promise<string | null>;
    onAppleLogin: () => Promise<string | null>;
    getIdToken: () => Promise<string | null>;
    loggedIn: () => Promise<boolean>;
    logout: () => Promise<void>;
  }
>;

const isStandaloneAndAndroid = () => {
  return Platform.OS === 'android' && Constants.appOwnership !== 'expo';
};

const Auth: FC<Props> = memo((props) => {
  const [state, setState] = useState<State>({
    email: '',
    uid: null,
    render: true,
    setup: false,
  });
  const isFirstRender = useIsFirstRender();
  const [syncCalendarsMutation] = useSyncCalendarsMutation({
    async onCompleted() {
      setRender();
    },
    onError(err) {
      Alert.alert('バックアップに失敗しました', err.message);
    },
  });

  const getIdToken = useCallback(async () => {
    return await auth.getIdToken();
  }, []);

  const setRender = useCallback(() => {
    setState((s) => ({
      ...s,
      render: true,
    }));
  }, []);

  const saveUser = useCallback(async () => {
    const idToken = await getIdToken();

    const response = await post('CreateUser', null, idToken || '');

    if (response?.error) {
      Alert.alert('ユーザーの保存に失敗しました。');
    }

    return Number(response?.status);
  }, [getIdToken]);

  const backupData = useCallback(async () => {
    const httpStatus = await saveUser();

    if (httpStatus === 201) {
      // ユーザーを新規で作成した場合はデータをサーバーに送る
      const { items, itemDetails, calendars } = await backup();

      const syncCalendars = calendars.map((v) => {
        const i = items.find((v1) => v1.id === v.itemId);
        let ids: SyncItemDetail[] = [];

        if (i) {
          ids = itemDetails
            .filter((v2) => v2.itemId === v.itemId)
            .map((id) => ({
              id: uuidv4(),
              title: id.title || '',
              kind: id?.kind || '',
              place: id?.place || '',
              url: id?.url || '',
              memo: id?.memo || '',
              priority: id?.priority || 1,
            }));
        }

        const r: SyncCalendar = {
          id: uuidv4(),
          date: dayjs(v.date).format('YYYY-MM-DDT00:00:00'),
          item: {
            id: uuidv4(),
            title: i?.title || '',
            kind: i?.kind || '',
            itemDetails: ids,
          },
        };

        return r;
      });

      const variables: SyncCalendarsMutationVariables = {
        calendars: {
          calendars: syncCalendars,
        },
      };

      syncCalendarsMutation({ variables });
    } else if (httpStatus === 200) {
      setRender();
    } else {
      Alert.alert('エラー', 'ユーザー登録/ログインに失敗しました', [
        {
          text: '閉じる',
          onPress: () => {
            setRender();
          },
        },
      ]);
    }
  }, [saveUser, syncCalendarsMutation, setRender]);

  const setSession = useCallback(
    async (refresh = false) => {
      const idToken = await auth.setSession(refresh);

      if (idToken) {
        const user = firebase.auth().currentUser;
        if (user) {
          setState((s) => ({
            ...s,
            render: false,
          }));
          await AsyncStorage.setItem('email', user?.email || '');
          await AsyncStorage.setItem('uid', user.uid);
          setState((s) => ({
            ...s,
            email: user.email || '',
            uid: user.uid || '',
          }));

          await backupData();
        }
      }

      return idToken;
    },
    [backupData]
  );

  const loggedIn = useCallback(async () => {
    const idToken = await getIdToken();

    return Boolean(idToken);
  }, [getIdToken]);

  const firebaseLogin = useCallback(
    async (credential: firebase.auth.OAuthCredential) => {
      const data = await firebase
        .auth()
        .signInWithCredential(credential)
        .catch((error: any) => {
          console.log(error);
        });
      console.log(data);

      return await setSession(true);
    },
    [setSession]
  );

  const firebaseGoogleLogin = useCallback(
    async (idToken: string, accessToken: string) => {
      const credential = firebase.auth.GoogleAuthProvider.credential(
        idToken,
        accessToken
      );

      await firebaseLogin(credential);
    },
    [firebaseLogin]
  );

  const onAppleLogin = useCallback(async () => {
    const nonceString = nonceGen(32);

    try {
      const result = await AppleAuthentication.signInAsync({
        requestedScopes: [
          AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
          AppleAuthentication.AppleAuthenticationScope.EMAIL,
        ],
        state: nonceString,
      });
      const provider = new firebase.auth.OAuthProvider('apple.com');
      const credential = provider.credential({
        idToken: result.identityToken || '',
        rawNonce: nonceString,
      });

      await firebaseLogin(credential);
      return await AsyncStorage.getItem('uid');
    } catch (e) {
      console.log(e);
      return null;
    }
  }, [firebaseLogin]);

  const onGoogleLogin = useCallback(async () => {
    if (isStandaloneAndAndroid()) {
      // TODO: AndroidのstandaloneのみGoogleSignInを使わないとエラーになる
      // https://github.com/expo/expo/issues/4762
      await GoogleSignIn.askForPlayServicesAsync();
      const result = await GoogleSignIn.signInAsync();

      if (result.type === 'success' && result.user && result.user.auth) {
        const { idToken, accessToken } = result.user.auth;
        await firebaseGoogleLogin(idToken || '', accessToken || '');
        return await AsyncStorage.getItem('uid');
      } else {
        Sentry.captureMessage(JSON.stringify(result));
        return null;
      }
    } else {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      const iosClientId = process.env.GOOGLE_LOGIN_IOS_CLIENT_ID;
      const result = await Google.logInAsync({
        clientId:
          Platform.OS === 'ios' ? String(iosClientId) : String(androidClientId),
        iosClientId,
        androidClientId,
        scopes: ['profile', 'email'],
      });

      console.log(result);

      if (result.type === 'success') {
        const { idToken, accessToken } = result;
        await firebaseGoogleLogin(idToken || '', accessToken || '');
        return await AsyncStorage.getItem('uid');
      } else {
        Sentry.captureMessage(JSON.stringify(result));
        return null;
      }
    }
  }, [firebaseGoogleLogin]);

  useEffect(() => {
    if (!isFirstRender) return;

    if (isStandaloneAndAndroid()) {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      try {
        GoogleSignIn.initAsync({
          clientId: String(androidClientId),
        });
      } catch ({ message }) {
        Sentry.captureMessage(JSON.stringify(message));
      }
    }

    const checkLogin = async () => {
      const login = await loggedIn();

      if (login && !state.uid) {
        const uid = await AsyncStorage.getItem('uid');

        if (uid) {
          setState((s) => ({
            ...s,
            uid,
            setup: true,
          }));
        }
      } else {
        setState((s) => ({
          ...s,
          setup: true,
        }));
      }

      if (login && !state.email) {
        const email = await AsyncStorage.getItem('email');
        if (email) {
          setState((s) => ({
            ...s,
            email,
          }));
        }
      }
    };

    firebase.auth().onAuthStateChanged(() => {
      checkLogin();
    });
  }, [isFirstRender, state, loggedIn]);

  const onLogout = useCallback(async () => {
    setState((s) => ({
      ...s,
      render: false,
    }));

    await logout();
    setState((s) => ({
      ...s,
      email: '',
      uid: null,
      render: true,
    }));
  }, []);

  if (!state.setup) {
    return null;
  }

  if (!state.render) {
    return (
      <View style={styles.container}>
        <Spinner
          visible={true}
          textContent="読込中"
          textStyle={{ color: theme().color.white }}
        />
      </View>
    );
  }

  return (
    <Provider
      value={{
        onGoogleLogin: onGoogleLogin,
        onAppleLogin: onAppleLogin,
        getIdToken: getIdToken,
        loggedIn: loggedIn,
        logout: onLogout,
        email: state.email,
        uid: state.uid,
      }}
    >
      {props.children}
    </Provider>
  );
});

const logout = async () => {
  await firebase.auth().signOut();

  await AsyncStorage.removeItem('id_token');
  await AsyncStorage.removeItem('expiration');
  await AsyncStorage.removeItem('email');
};

export const Consumer = Context.Consumer;
export const useAuth = () => useContext(Context);

export default Auth;

const nonceGen = (length: number) => {
  let result = '';
  let characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};

const styles = EStyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: '$background',
    height: '100%',
  },
});
