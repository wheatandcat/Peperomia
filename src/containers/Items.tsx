import React, {
  FC,
  createContext,
  memo,
  useState,
  useCallback,
  useContext,
} from 'react';
import { SelectItem } from 'domain/item';
import { SelectCalendar } from 'domain/calendar';
import { SelectItemDetail } from 'domain/itemDetail';
import { getItems } from 'lib/item';
import { getItemDetails } from 'lib/itemDetail';
import { getCalendars } from 'lib/calendar';
import { useDidMount } from 'hooks/index';
import { UID } from 'domain/user';
import { Context as AuthContext } from './Auth';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ItemAbout = {
  itemId: number | string;
  about: string;
};

type Props = {};

type State = {
  loading: boolean;
  items: SelectItem[];
  itemDetails: SelectItemDetail[];
  about: ItemAbout[];
  calendars: SelectCalendar[];
};

export type ContextProps = Partial<
  Pick<State, 'items' | 'itemDetails' | 'about' | 'calendars'> & {
    refreshData: (targetUid?: UID | false) => void;
    itemsLoading: boolean;
  }
>;

const initialState = () => ({
  loading: true,
  calendarsLoading: true,
  items: [],
  itemDetails: [],
  about: [],
  calendars: [],
});

const Connected: FC<Props> = memo((props) => {
  const [state, setState] = useState<State>(initialState());
  const { uid } = useContext(AuthContext);

  const setItemsDetail = useCallback(
    (data: SelectItemDetail[]) => {
      const names = data.map((val: SelectItemDetail) => val.title).join(' â†’ ');
      const itemId = data[0].itemId;

      const about = [
        ...state.about,
        {
          itemId,
          about: names,
        },
      ];

      setState((s) => ({
        ...s,
        about,
        loading: false,
        itemDetails: [...state.itemDetails, ...data],
      }));
    },
    [state]
  );

  const getData = useCallback(
    async (targetUid: UID | false = false) => {
      setState((s) => ({
        ...s,
        loading: true,
      }));

      let uid2 = targetUid === false ? uid : targetUid;

      const items = await getItems(uid2);

      setState((s) => ({
        ...s,
        items: items,
        about: [],
      }));

      items.map(async (val) => {
        const itemDetails = await getItemDetails(uid2, String(val.id));

        if (!itemDetails || itemDetails.length === 0) {
          return;
        }

        setItemsDetail(itemDetails);
      });

      const calendars = await getCalendars(uid2);

      setState((s) => ({
        ...s,
        calendars,
        loading: false,
      }));
    },
    [setItemsDetail, uid]
  );

  useDidMount(() => {
    getData();
  });

  return (
    <Provider
      value={{
        items: state.items,
        itemDetails: state.itemDetails,
        calendars: state.calendars,
        about: state.about,
        refreshData: getData,
        itemsLoading: state.loading,
      }}
    >
      {props.children}
    </Provider>
  );
});

export default Connected;

export const useItems = () => useContext(Context);
export const Consumer = Context.Consumer;
