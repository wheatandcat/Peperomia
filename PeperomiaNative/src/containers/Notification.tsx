import React, { memo, createContext, FC, useCallback, useContext } from 'react';
import { Platform, Alert } from 'react-native';
import * as Notifications from 'expo-notifications';
import * as Permissions from 'expo-permissions';
import Constants from 'expo-constants';
import { useFetch } from 'containers/Fetch';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

export type ContextProps = Partial<{
  onPermissionRequest: () => Promise<boolean>;
}>;

type Props = {};

const Notification: FC<Props> = memo((props) => {
  const { post } = useFetch();

  const onPermissionRequest = useCallback(async () => {
    if (!post) {
      return false;
    }

    if (!Constants.isDevice) {
      Alert.alert('端末から実行してくだださい');
      return false;
    }

    const { status: existingStatus } = await Permissions.getAsync(
      Permissions.NOTIFICATIONS
    );

    let finalStatus = existingStatus;
    if (existingStatus !== Permissions.PermissionStatus.GRANTED) {
      const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);
      finalStatus = status;
    }

    if (finalStatus !== 'granted') {
      Alert.alert('Push通知のトークンの取得に失敗しました');
      return false;
    }

    const { data: token } = await Notifications.getExpoPushTokenAsync();

    if (Platform.OS === 'android') {
      Notifications.setNotificationChannelAsync('default', {
        name: 'default',
        importance: Notifications.AndroidImportance.MAX,
        vibrationPattern: [0, 250, 250, 250],
        lightColor: '#FF231F7C',
      });
    }

    const request = {
      pushToken: {
        token,
        deviceId: Constants.deviceId,
      },
    };

    //Alert.alert(token);
    //console.log(token);

    const response = await post('CreatePushToken', request);

    if (response.error) {
      Alert.alert('更新に失敗しました');
      return false;
    }

    return true;
  }, [post]);

  return <Provider value={{ onPermissionRequest }}>{props.children}</Provider>;
});

export default Notification;

export const useNotification = () => useContext(Context);
export const Consumer = Context.Consumer;
