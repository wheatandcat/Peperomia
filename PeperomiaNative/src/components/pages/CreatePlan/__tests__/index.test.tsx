import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import * as Items from 'containers/Items';
import * as Auth from 'containers/Auth';
import Index from '../index';

describe('components/pages/CreatePlan/index.tsx', () => {
  let wrapper: ShallowWrapper;

  const propsData: any = () => ({
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        date: '1',
        kind: 'park',
      },
    },
  });

  beforeEach(() => {
    jest.spyOn(Items, 'useItems').mockImplementation(
      () =>
        ({
          refreshData: jest.fn(),
          items: [],
          calendars: [],
        } as any)
    );
    jest.spyOn(Auth, 'useAuth').mockImplementation(
      () =>
        ({
          uid: 'test',
        } as any)
    );

    wrapper = shallow(<Index {...propsData()} />);
  });

  it('正常に表示されている', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
