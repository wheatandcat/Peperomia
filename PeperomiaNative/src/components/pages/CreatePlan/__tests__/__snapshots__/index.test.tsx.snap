// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/CreatePlan/index.tsx 正常に表示されている 1`] = `
<Memo()
  calendars={Array []}
  items={Array []}
  navigation={
    Object {
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  refreshData={[MockFunction]}
  route={
    Object {
      "params": Object {
        "date": "1",
        "kind": "park",
      },
    }
  }
  uid="test"
/>
`;
